internal/porchutil/function.go

```diff
@@ -0,0 +1,155 @@
+package porchutil
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/GoogleContainerTools/kpt/porch/api/porch/v1alpha1"
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/client-go/kubernetes"
+
+	// Enable the GCP Authentication plugin
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const expirationNanoSecond = 3
+
+var (
+	functionAPI     = "/apis/" + v1alpha1.SchemeGroupVersion.String() + "/functions"
+	PorchExpiration = time.Duration(expirationNanoSecond) * time.Second
+)
+
+// NewPorchFunctionClient provides the PorchFunctionClient which connects to Porch server
+// and gets the functions.
+func NewPorchFunctionClient() *PorchFunctionClient {
+	functions, err := visitFunctions()
+	if err != nil {
+		return &PorchFunctionClient{IsPorchAvailable: false, Functions: nil}
+	}
+	return &PorchFunctionClient{IsPorchAvailable: true, Functions: functions}
+}
+
+type PorchFunctionClient struct {
+	IsPorchAvailable bool
+	Functions        []v1alpha1.Function
+}
```

Comment:
slight conceptual misalignment. this struct is actually not a client. I wonder if it would work as well to have a simpler interface, possibly avoiding the IsPorchAvailable bool and return list of functions directly?

(I am not feeling too strongly though)

Reply:
that's true. I defined the struct to make the caller code be more specific that only when porch available will we use  porch functions
```
if client.IsPorchAvailable():
   client.ListFunctions
```
I don't have a strong preference but agree the naming is misaligned. I'll update.

Reply:
set timeout to 10.